### PR TITLE
[Issue: 2822] Metadata import email parameter changes

### LINKED ISSUE #2822: "metadata-import" script should NOT allow email to be specified via REST API /scripts endpoint
Created based on feedback in https://github.com/DSpace/dspace-angular/pull/636#pullrequestreview-441228120

Currently, in the Scripts & Processes UI, any administrator can specify any `--email` param for the `metadata-import` script. This is improper behavior as  any Administrator can batch edit metadata as if they were anyone else -- this would bypass all normal authorization checks within DSpace, so it is a minor security issue (even though Administrators are trusted, they should not be able to bypass permissions this easily).

Ideally, the REST API should NOT allow a `--email` parameter to be passed in.  Instead, it should be automatically set to the email of the currently authenticated user. This also ensures that logs detail exactly which user kicked off the import script from the Admin UI.

### PR DESCRIPTION
**Pull request by Atmire-github**
## References
* Fixes [GitHub issue](https://github.com/DSpace/DSpace/issues/2822)

## Description
This PR will remove the "email" option from the REST metadata-import script. The email parameter is still available (& mandatory) for the CLI implementation of this script. The REST implementation of the metadata-import script will use the currently logged in user to execute the script.

## Instructions for Reviewers

Retrieve the configuration of the REST metadata-import script, the email option should not be available here. In the CLI the option will be present & will be mandatory.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [X] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [X] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

**Comment by tdonohue** [comment]
@Atmire-github or @benbosman : This looks to be failing some tests.  Likely not all tests were updated to support this yet.

**Comment by KevinVdV** [comment]
@tdonohue FYI: This PR was fired by me (I was accidentally logged in as another user), I have delegated a fix for this to Raf, I will ping you when it is ready.

**Comment by KevinVdV** [comment]
@tdonohue Raf merged the latest master into this branch & fixed the tests.

File: dspace-api/src/test/java/org/dspace/app/csv/CSVMetadataImportReferenceIT.java [hunk 1]
```diff
@@ -19,54 +19,132 @@
...
      */
     @Before
-    public void setup() {
+    public void setup() throws SQLException, AuthorizeException {
+        context.turnOffAuthorisationSystem();
+        parentCommunity = communityService.create(null, context);
+        communityService.addMetadata(context, parentCommunity, "dc", "title", null, null, "Parent Community");
+
+        col1 = collectionService.create(context, parentCommunity);
+        collectionService.addMetadata(context, col1, "dc", "title", null, null, "Collection 1");
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/app/csv/CSVMetadataImportReferenceIT.java:86]: We should be using `Builder` classes in this `setup()` method to create *all* new content.  That would mean that we can completely *remove* the `destroy()` class (as Builders cleanup after themselves). 

It looks like the old code in this `setup()` used Builders, but you've replaced/removed that code.  Please revert back to using `Builders` again to create the Community, Collection (see above) and all Entities (see below). 

**If you rebase on `main`, the Builder classes are now available in `dspace-api`.**

File: dspace-api/src/test/java/org/dspace/app/csv/CSVMetadataImportReferenceIT.java [hunk 2]
```diff
@@ -19,54 +19,132 @@
...
+                    null, 0, null, false, true);
+
+        context.restoreAuthSystemState();
+
         context.restoreAuthSystemState();
     }
 
+    @After
+    @Override
+    public void destroy() {
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/app/csv/CSVMetadataImportReferenceIT.java:113]: Once you are using Builders, this entire `destroy()` method can be removed.

File: dspace-api/src/test/java/org/dspace/app/csv/CSVMetadataImportReferenceIT.java [hunk 3]
```diff
@@ -175,12 +253,13 @@ public void testMultiRowNameRef() throws Exception {
     @Test
     public void testSingleUUIDReference() throws Exception {
         context.turnOffAuthorisationSystem();
-        Item person = ItemBuilder.createItem(context, col1)
-                                 .withRelationshipType("Person")
-                                 .build();
+        WorkspaceItem workspaceItem = workspaceItemService.create(context, col1, false);
+        Item person = installItemService.installItem(context, workspaceItem);
+        itemService.addMetadata(context, person, "relationship", "type", null, null, "Person");
+        itemService.update(context, person);
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/app/csv/CSVMetadataImportReferenceIT.java:259]: Again, remove this new code and revert to using the `ItemBuilder` class.  We require Builders.

File: dspace-api/src/test/java/org/dspace/app/csv/CSVMetadataImportReferenceIT.java [hunk 4]
```diff
@@ -192,12 +271,14 @@ public void testSingleUUIDReference() throws Exception {
...
-                                  .withRelationshipType("Person")
-                                  .build();
+        WorkspaceItem workspaceItem = workspaceItemService.create(context, col1, false);
+        Item person = installItemService.installItem(context, workspaceItem);
+        itemService.addMetadata(context, person, "relationship", "type", null, null, "Person");
+        itemService.update(context, person);
+        WorkspaceItem workspaceItem2 = workspaceItemService.create(context, col1, false);
+        Item person2 = installItemService.installItem(context, workspaceItem2);
+        itemService.addMetadata(context, person2, "relationship", "type", null, null, "Person");
+        itemService.update(context, person2);
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/app/csv/CSVMetadataImportReferenceIT.java:281]: Same here, revert to using Builders again

File: dspace-api/src/test/java/org/dspace/app/csv/CSVMetadataImportReferenceIT.java [hunk 5]
```diff
@@ -214,14 +295,16 @@ public void testMultiUUIDReference() throws Exception {
...
         context.turnOffAuthorisationSystem();
-        Item person = ItemBuilder.createItem(context, col1)
-                                 .withTitle("Person")
-                                 .withRelationshipType("Person")
-                                 .build();
+        WorkspaceItem workspaceItem = workspaceItemService.create(context, col1, false);
+        Item person = installItemService.installItem(context, workspaceItem);
+        itemService.addMetadata(context, person, "relationship", "type", null, null, "Person");
+        itemService.addMetadata(context, person, "dc", "title", null, null, "Person");
+        itemService.update(context, person);
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/app/csv/CSVMetadataImportReferenceIT.java:302]: Same here, revert to using Builders again

File: dspace-api/src/test/java/org/dspace/app/csv/CSVMetadataImportReferenceIT.java [hunk 6]
```diff
@@ -236,18 +319,21 @@ public void testMultiRefArchivedCsv() throws Exception {
...
+        WorkspaceItem workspaceItem2 = workspaceItemService.create(context, col1, false);
+        Item person2 = installItemService.installItem(context, workspaceItem2);
+        itemService.addMetadata(context, person2, "relationship", "type", null, null, "Person");
+        itemService.addMetadata(context, person2, "dc", "title", null, null, "Person2");
+        itemService.update(context, person2);
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/app/csv/CSVMetadataImportReferenceIT.java:331]: Same here, revert to using Builders again

File: dspace-api/src/test/java/org/dspace/app/csv/CSVMetadataImportReferenceIT.java [hunk 7]
```diff
@@ -300,14 +386,17 @@ public void testNonUniqueRowName() throws Exception {
...
+        WorkspaceItem workspaceItem = workspaceItemService.create(context, col1, false);
+        Item person = installItemService.installItem(context, workspaceItem);
+        itemService.addMetadata(context, person, "relationship", "type", null, null, "Person");
+        itemService.addMetadata(context, person, "dc", "identifier", "other", null, "1");
+        itemService.update(context, person);
+        WorkspaceItem workspaceItem2 = workspaceItemService.create(context, col1, false);
+        Item person2 = installItemService.installItem(context, workspaceItem2);
+        itemService.addMetadata(context, person2, "relationship", "type", null, null, "Person");
+        itemService.addMetadata(context, person2, "dc", "identifier", "other", null, "1");
+        itemService.update(context, person2);
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/app/csv/CSVMetadataImportReferenceIT.java:398]: Same here, revert to using Builders again.  I'll stop repeating this message now.  Essentially, **MOST** of your changes to this IT class can now be reverted. You can use Builders here as the Builder classes are now available in `dspace-api`.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CsvImportIT.java [hunk 8]
```diff
@@ -222,22 +240,101 @@ private Item validateSpecificItemRelationCreationCsvImportMultiple(Collection co
...
+        } finally {
+            ProcessBuilder.deleteProcess(idRef.get());
         }
-        out.flush();
-        out.close();
-        out = null;
+    }
+
+    @Test
+    public void createRelationshipsWithCsvImportWithSpecifiedEPersonParameterTest() throws Exception {
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CsvImportIT.java:279]: We may want to move this test to `dspace-api`, as it looks like it's testing the *command-line* script can still use `-e`.  It's a good test, but it doesn't belong in the `dspace-server-webapp`, as the `dspace-server-webapp` is NOT allowed to use `-e`.  So, I think we should have a test in this class that proves the REST endpoint **cannot** use `-e` but still succeeds without it.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@KevinVdV and @Atmire-github :  Overall the code here looks good & it works. I've tested it with the UI as well and the `-e` param is no longer available to http://localhost:4000/processes/new?script=metadata-import  So, that's all great.

However, the ITs here need some work.  As you may have seen, we recently moved the Builders over to `dspace-api` in #2897.  It looks like this branch needs to be rebased on latest `main`, so that you can also use those Builders in the tests you created here (most of your tests look to not be using Builders which makes them a lot more complex & requires a lot of manual cleanup).  Please update your ITs to use Builders, and then I'm OK with this PR as-is.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CsvImportIT.java (hunk 8 above)
Comment (KevinVdV) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/csv/CsvImportIT.java]: @tim-atmire There might be a bit confusion about what this test does. This test checks that if you provide the "-e" option to the metadata-import script from the REST API that the script rejects this.

To make it a bit more clear, we made the following alterations:
* We renamed the method
* We don't just check that the item wasn't imported, but also check that the process has failed due to the invalid option.

**Comment by KevinVdV** [comment]
@tdonohue We have merged the latest master into this working branch & used the builders in the dspace-api.

**Review by tdonohue — APPROVED** [review]
👍 This looks good to me now.  ITs now all look good.  Thanks for the updates & explanation @KevinVdV !

**Comment by tdonohue** [comment]
Merging, as this was previously flagged as `1 Approval`

**Pull request merged**